lib/kubernetes-deploy/runner_task.rb

```diff
@@ -8,51 +8,99 @@ module KubernetesDeploy
   class RunnerTask
     include KubeclientBuilder
 
-    class FatalTaskRunError < FatalDeploymentError; end
-    class TaskTemplateMissingError < FatalDeploymentError
-      def initialize(task_template, namespace, context)
-        super("Pod template `#{task_template}` cannot be found in namespace: `#{namespace}`, context: `#{context}`")
-      end
-    end
+    class TaskConfigurationError < FatalDeploymentError; end
+    class TaskTemplateMissingError < TaskConfigurationError; end
+
+    attr_reader :pod_name
 
-    def initialize(namespace:, context:, logger:)
+    def initialize(namespace:, context:, logger:, max_watch_seconds: nil)
       @logger = logger
       @namespace = namespace
-      @kubeclient = build_v1_kubeclient(context)
       @context = context
+      @max_watch_seconds = max_watch_seconds
     end
 
     def run(*args)
       run!(*args)
       true
-    rescue FatalDeploymentError => error
-      @logger.fatal "#{error.class}: #{error.message}"
+    rescue DeploymentTimeoutError, FatalDeploymentError
       false
     end
 
-    def run!(task_template:, entrypoint:, args:, env_vars: [])
+    def run!(task_template:, entrypoint:, args:, env_vars: [], verify_result: true)
       @logger.reset
-      @logger.phase_heading("Validating configuration")
+
+      @logger.phase_heading("Initializing task")
       validate_configuration(task_template, args)
-      if kubectl.server_version < Gem::Version.new(MIN_KUBE_VERSION)
-        @logger.warn(KubernetesDeploy::Errors.server_version_warning(kubectl.server_version))
+      pod = build_pod(task_template, entrypoint, args, env_vars, verify_result)
+      validate_pod(pod)
+
+      @logger.phase_heading("Running pod")
+      create_pod(pod)
+
+      if verify_result
+        @logger.phase_heading("Streaming logs")
+        watch_pod(pod)
+      else
+        record_status_once(pod)
       end
-      @logger.phase_heading("Fetching task template")
-      raw_template = get_template(task_template)
+      @logger.print_summary(:success)
+    rescue DeploymentTimeoutError
+      @logger.print_summary(:timed_out)
+      raise
+    rescue FatalDeploymentError
+      @logger.print_summary(:failure)
+      raise
+    end
 
-      @logger.phase_heading("Constructing final pod specification")
-      rendered_template = build_pod_template(raw_template, entrypoint, args, env_vars)
+    private
 
-      validate_pod_spec(rendered_template)
+    def create_pod(pod)
+      @logger.info "Creating pod '#{pod.name}'"
+      pod.deploy_started_at = Time.now.utc
+      kubeclient.create_pod(Kubeclient::Resource.new(pod.definition))
+      @pod_name = pod.name
```

Comment:
Why an instance var?

Reply:
Because today, the pod's name is exposed as the return value of `run`. I don't want to keep doing that, but there are legitimate use cases for wanting to know the name of the pod created so you can take follow-up action. For example, the caller might want to rescue DeploymentTimeoutError, delete the pod created (to make sure two aren't running at once), and retry the runner. Incidentally, I'm wondering if we should add a `--delete-pod-on-timeout` option to the runner itself (not in this PR).